### PR TITLE
feat(Core/Instance): Add instance validation for creature scripts

### PR DESCRIPTION
**Pull request by Kitzunu**
Co-Authored-By: Shauren <shauren.trinity@gmail.com>

https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/CreatureAIImpl.h
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Instances/InstanceScript.cpp#L810-L816

## Changes Proposed:
-  Add validation for creature scripts in instances. Don't initialize the script if the creature is not in the instance.

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/4556

## Tests Performed:
- Builds
- Tested in-game

## How to Test the Changes:

**FIRST OF ALL: RUN CMAKE!**
#### Test Case 1
1. Spawn a creature outside the instance.
2. for example: `.npc add temp 16808` 
3. Engage in combat, see that he is not yelling anything.
4. for example: `.npc add temp 38153`
5. See that the server does not crash =)

#### Test Case 2
1. Choose an instance, and see that the scripts work as they should.
2. for example `.go c id 16808`
3. Engage in combat, see that he is yelling.

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
@Kitzunu the build is still failing, please fix it before it can be tested

**Comment by Kitzunu** [comment]
Q.q fuck ubuntu
I can look at it on Saturday

**Comment by Kitzunu** [comment]
I can't understand how sObjectMgr can be temporary?

**Comment by Kitzunu** [comment]
Somebody, help pls D:

i_script_id is considered temporary
```cpp
std::string const& InstanceMap::GetScriptName() const
{
    return sObjectMgr->GetScriptName(i_script_id);
}
```
Probably due to that it is converting uint32 to std::string in a way that isn't good, but how could I make it good?

**Comment by Kitzunu** [comment]
This is the only thing that we need to fix because the PR works as it should.
All we need to do is make ubuntu happy q.q

**Comment by FrancescoBorzi** [comment]
> This is the only thing that we need to fix because the PR works as it should.
> All we need to do is make ubuntu happy q.q

I can't find that piece of code in the error, instead I see another error:

```
/home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/server/game/Maps/Map.cpp:2821:5: fatal error: use of undeclared identifier 'instance_script'
    instance_script = sEluna->GetInstanceData(this);
```

**Comment by BarbzYHOOL** [comment]
"uint32 to std::string"
why would it do that? (sry didn't check the code but it sounds weird)

**Comment by Kitzunu** [comment]
No maybe I was wrong, I can't check now, but GetScriptName probably returns char

**Comment by Kitzunu** [comment]
Not gonna resolve the conflicts of 360 files before I get the build to pass

**Comment by Kitzunu** [comment]
Nice all the builds that matters are passing :)
And I am very aware Eluna is failing, but that is up to the maintainer of the module to fix. Not me.

Ready to test and review :)

**Comment by FrancescoBorzi** [comment]
@Rochet2 @ayasecore any hint about how to fix Eluna here? the module build is failing with:

```
src/server/game/Maps/Map.cpp:2821 
 fatal error: use of undeclared identifier 'instance_script'
    instance_script = sEluna->GetInstanceData(this);
````

**Comment by Kitzunu** [comment]
instance_script -> i_data

**Comment by LightDestory** [comment]
I am going to test this PR soon, I will update asap

**Comment by LightDestory** [comment]
Confimed, the server doesn't crash anymore.

**Comment by FrancescoBorzi** [comment]
@Kitzunu @LightDestory apart from testing that the original bug has actually been fixed, have you guys also checked instances in general to see that everything is in place? as this is quite a big change and we should verify it before merging

**Comment by Kitzunu** [comment]
image

Now I totally agree with you @FrancescoBorzi 
Perhaps @pklloveyou could put it on his server for a few days just to make sure that nothing is out of the ordinary :)

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers  @LightDestory can you please test some instances with this PR? in order to make sure there is no regression.

**Comment by LightDestory** [comment]
I don't have a huge knowledge of WoW instances, I don't know their ids neither. Moreover right now I am busy with university lessons so get some spare time for this is a problem for me. I am sorry.

**Comment by Kitzunu** [comment]
LF review q.q

**Comment by FrancescoBorzi** [comment]
@Yehonal @azerothcore/testers @azerothcore/developers @pklloveyou @mpfans can you please review/test this?

**Comment by claudiodfc** [comment]
#### Here's my procedure on MSVC2019 (Windows 10)
- generated new solution using cmake
- clean build
- build solution

#### Results
- [X] NPC 16808 didn't yell anything out of instance
- [X] NPC 38153 didn't crash the server out of instance
- [X] NPC 16808 working script when instanced

**Comment by claudiodfc** [comment]
Hey @Kitzunu , Codefactor is throwing this warning:

`Consider using rand_r(...) instead of rand(...) for improved thread safety. `

**Comment by Kitzunu** [comment]
Yea, ignore it

File: src/server/game/AI/CreatureAIImpl.h
Comment (FrancescoBorzi) [inline src/server/game/AI/CreatureAIImpl.h:66]:
```suggestion
return InstanceHasScript(obj, scriptName) ? new AI(obj) : nullptr
```

File: src/server/game/Maps/Map.h
Comment (FrancescoBorzi) [inline src/server/game/Maps/Map.h:691]: please use `[[nodiscard]]` on all getters that don't cause any side-effect

Comment (claudiodfc) [reply src/server/game/Maps/Map.h:691]: I was reading about it but can't understand what is this `[[nodiscard]]`

Comment (FrancescoBorzi) [reply src/server/game/Maps/Map.h:691]: @claudiodfc C++17 feature, basically marking a method as `[[nodiscard]]` will make the compiler throw an error automatically whenever the result of that method is ignored. So it's useful to mark getter-methods that are supposed to just give you some data without causing any side effect, because whenever you ignore the result of such a method, you are probably doing something wrong ;)

Comment (claudiodfc) [reply src/server/game/Maps/Map.h:691]: Got it, thanks!

**Comment by Kitzunu** [comment]
Updated

File: src/server/game/AI/CreatureAIImpl.h
Comment (FrancescoBorzi) [inline src/server/game/AI/CreatureAIImpl.h:61]: indentation

File: src/server/scripts/Northrend/Ulduar/Ulduar/ulduar.h
Comment (Winfidonarleyan) [inline src/server/scripts/Northrend/Ulduar/Ulduar/ulduar.h:8]: dobule include

File: src/server/scripts/Outland/CoilfangReservoir/SlavePens/boss_ahune.cpp
Comment (Winfidonarleyan) [inline src/server/scripts/Outland/CoilfangReservoir/SlavePens/boss_ahune.cpp:12]: maybe move to first include?

File: src/server/scripts/Outland/BlackTemple/black_temple.h
Comment (Winfidonarleyan) [inline src/server/scripts/Outland/BlackTemple/black_temple.h:10]: double include

File: src/server/scripts/Outland/BlackTemple/black_temple.h
Comment (Winfidonarleyan) [inline src/server/scripts/Outland/BlackTemple/black_temple.h:11]: double include

File: src/server/scripts/Outland/Auchindoun/ManaTombs/boss_nexusprince_shaffar.cpp
Comment (Winfidonarleyan) [inline src/server/scripts/Outland/Auchindoun/ManaTombs/boss_nexusprince_shaffar.cpp:8]: maybe move to first?

**Review by Winfidonarleyan — APPROVED** [review]

File: src/server/game/AI/CreatureAIImpl.h
Comment (Winfidonarleyan) [reply src/server/game/AI/CreatureAIImpl.h:61]: add 4 whitespace and `;` in end

**Comment by FrancescoBorzi** [comment]
@Kitzunu build still failing

**Comment by Kitzunu** [comment]
Due to one of the suggestions... I can revert them in the week

File: src/server/game/AI/CreatureAIImpl.h
Comment (Kitzunu) [reply src/server/game/AI/CreatureAIImpl.h:61]: https://github.com/azerothcore/azerothcore-wotlk/pull/4596#discussion_r604199917 I blame shin D: :P

File: src/server/scripts/Outland/CoilfangReservoir/SlavePens/boss_ahune.cpp
Comment (Kitzunu) [reply src/server/scripts/Outland/CoilfangReservoir/SlavePens/boss_ahune.cpp:12]: alphabetical

File: src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h
Comment (pklloveyou) [inline src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: This is an error. I'm sorry I just noticed this.

Comment (pklloveyou) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: https://github.com/Kitzunu/azerothcore-wotlk/blob/058aea0652d40241704db91c5a177f63f4b59d32/src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/instance_hellfire_ramparts.cpp#L12

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/utgarde_pinnacle.h
Comment (pklloveyou) [inline src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/utgarde_pinnacle.h:11]: It's the same here

Comment (pklloveyou) [reply src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/utgarde_pinnacle.h:11]: https://github.com/Kitzunu/azerothcore-wotlk/blob/058aea0652d40241704db91c5a177f63f4b59d32/src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/instance_utgarde_pinnacle.cpp#L12

File: src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h
Comment (Kitzunu) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: I don't understand when you just link a random line.

I don't think this is an error as I just define the script name for the instance.
This means that if someone in the future decides to script the creatures in C++ they have this already made.

Comment (pklloveyou) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: The bio code script in the instance does not work

Comment (Kitzunu) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: Im sorry for not understanding.
Give me some steps to test and I can check tomorrow

Comment (pklloveyou) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: Enter the instance and fight with the boss. You will find that they are just like outside the instance. The script doesn't work

Comment (pklloveyou) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: Change the line to
```c++
 #define HellfireRampartsScriptName "instance_ramparts"
 ```
Fight again, and you'll find the script works

File: src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/utgarde_pinnacle.h
Comment (pklloveyou) [reply src/server/scripts/Northrend/UtgardeKeep/UtgardePinnacle/utgarde_pinnacle.h:11]:
```c++
#define UtgardePinnacleScriptName "instance_pinnacle"
```

File: src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h
Comment (Kitzunu) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]: Oh now I get it.
Tbh we should change it to instance_hellfire_ramparts in the scropt file.

I will do this tomorrow, good catch 👍

Comment (pklloveyou) [reply src/server/scripts/Outland/HellfireCitadel/HellfireRamparts/hellfire_ramparts.h:11]:
```sql
UPDATE `instance_template` SET  script = 'instance_hellfire_ramparts' WHERE `map` = 543;
UPDATE `instance_template` SET  script = 'instance_utgarde_pinnacle' WHERE `map` = 575;
```

**Pull request merged**
